ci(cspell-bot): group words by file to reduce log size

OnkarRuikar opened the pull request:
## Problem

In the weekly [spelling check bot](https://github.com/mdn/content/blob/main/.github/workflows/spelling-check-bot.yml#L1), the current scheme of reporting each word on a separate line with corresponding file is generating massive logs. If someone adds a different version of lorem ipsum then it breaks the GitHub workflow with an error 

> GraphQL: Body is too long (maximum is 65536 characters) (createIssue)

https://github.com/mdn/content/actions/runs/13753835872/job/38458187038#step:5:876

> ![1](https://github.com/user-attachments/assets/47eb3973-a2a0-4b67-a0d5-6e46d54c992d)

## Solution

Instead of having file-word combinations, have file-word**s** combinations. That is, list a file path followed by the list of unknown words in it. It will reduce `files` x `words` lines to only `files`.

For example, the equivalent outcome of the failed workflow would be:

> - [files/en-us/learn_web_development/howto/solve_html_problems/use_javascript_within_a_webpage/index.md](https://github.com/mdn/content/blob/main/files/en-us/learn_web_development/howto/solve_html_problems/use_javascript_within_a_webpage/index.md?plain=1) - Unknown words (Matuzovic)
> - [files/en-us/web/api/performance_api/monitoring_bfcache_blocking_reasons/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/api/performance_api/monitoring_bfcache_blocking_reasons/index.md?plain=1) - Unknown words (idbversionchangeevent, idledetector, keyboardlock, otpcredential, pictureinpicturewindow, smartcardconnection, speechrecognition, webhid, webxrdevice)
> - [files/en-us/web/api/trusted_types_api/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/api/trusted_types_api/index.md?plain=1) - Unknown words (tinyfill)
> - [files/en-us/web/css/_colon_first-child/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/_colon_first-child/index.md?plain=1) - Unknown words (Adhemar, Blankers, Bubka, Gebrselassie, Haile, Irena, Junxia, Kersee, Koen, Paavo, Szewinska, Usain, Zatopek)
> - [files/en-us/web/css/_colon_lang/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/_colon_lang/index.md?plain=1) - Unknown words (czaszką, lepiej, napisem, narysowaną, wchodzić, widzimy, znak, środka)
> - [files/en-us/web/css/_colon_last-child/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/_colon_last-child/index.md?plain=1) - Unknown words (Adhemar, Blankers, Bubka, Gebrselassie, Haile, Irena, Junxia, Kersee, Koen, Paavo, Szewinska, Usain, Zatopek)
> - [files/en-us/web/css/_colon_nth-child/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/_colon_nth-child/index.md?plain=1) - Unknown words (Adhemar, Blankers, Bubka, Gebrselassie, Haile, Irena, Junxia, Kersee, Koen, Paavo, Szewinska, Usain, Zatopek)
> - [files/en-us/web/css/_colon_nth-last-child/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/_colon_nth-last-child/index.md?plain=1) - Unknown words (Cloquet, Hinckley, Kursha, Miramichi, Peshtigo)
> - [files/en-us/web/css/_colon_only-child/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/_colon_only-child/index.md?plain=1) - Unknown words (Downey, Scarlett)
> - [files/en-us/web/css/_doublecolon_after/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/_doublecolon_after/index.md?plain=1) - Unknown words (Indo, scorpionfish)
> - [files/en-us/web/css/_doublecolon_first-letter/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/_doublecolon_first-letter/index.md?plain=1) - Unknown words (Robison, dragonfish)
> - [files/en-us/web/css/_doublecolon_first-line/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/_doublecolon_first-line/index.md?plain=1) - Unknown words (Exocoetidae)
> - [files/en-us/web/css/_doublecolon_marker/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/_doublecolon_marker/index.md?plain=1) - Unknown words (Deke, Gordo, Grissom, Schirra, Slayton)
> - [files/en-us/web/css/animation/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/animation/index.md?plain=1) - Unknown words (slidein)
> - [files/en-us/web/css/clip-path/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/clip-path/index.md?plain=1) - Unknown words (Aéro)
> - [files/en-us/web/css/counter/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/counter/index.md?plain=1) - Unknown words (Bron, Dwyane, Giancarlo, Iniesta, Xavi)
> - [files/en-us/web/css/css_conditional_rules/container_scroll-state_queries/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/css_conditional_rules/container_scroll-state_queries/index.md?plain=1) - Unknown words (Aristonem, Aristoxeni, Atqui, Beatus, Conferam, Duarum, Falli, Huius, Lyco, Memini, Metrodori, Nescio, Neutrum, Omnes, Omnia, Pisonem, Polemonis, Praeteritis, Prioris, Putabam, Quonam, Rationis, Tria, Triarium, Videmus, abducam, adesse, aequius, agas, altera, angustiarum, apertissimis, apparet, auctoritatem, beate, bonorum, capienda, causa, censent, commendas, concedamus, conquiescere, consumptum, controversia, controversiam, cuique, deinde, dicitis, dicitur, dicuntur, differt, discipulum, dissensio, dissensione, disseramus, dissimile, dixisse, dixisset, docilitas, eiusdem, equidem, erit, erunt, expositum, faciam, faciat, falsone, finem, fingunt, futtiles, gaudeo, haec, hilaretur, homines, ielunior, igitur, ille, infantes, ingenium, inprudens, instituta, ipsis, istarum, iucundum, iudicare, iustitia, laboro, laudatur, leves, liberius, liberos, locuples, longi, loqui, maior, maximarum, melius, memoria, motum, musicis, natos, nimium, nomine, occurrentia, oculis, omnes, oportere, oportunitatis, oratione, paria, peccata, pecuniae, perfectio, pertinacem, philosophiae, philosophos, possint, possumus, potest, potestne, praetervolavit, privasse, pulsi, putas, quaerere, recurrant, redargueret, reperies, ridens, saepius, satis, satisfacit, sensus, sequens, sequuntur, sordidos, subicias, subtilius, sumus, superiori, tantae, targetting, timore, utebare, utilitatem, vanos, veniunt, vere, videmus, vides, vincunt, vitarum, vivere)
> - [files/en-us/web/css/font-optical-sizing/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/font-optical-sizing/index.md?plain=1) - Unknown words (slablike)
> - [files/en-us/web/css/shape-image-threshold/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/shape-image-threshold/index.md?plain=1) - Unknown words (Aéro)
> - [files/en-us/web/css/shape-margin/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/shape-margin/index.md?plain=1) - Unknown words (Aéro)
> - [files/en-us/web/css/shape-outside/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/shape-outside/index.md?plain=1) - Unknown words (Aéro)
> - [files/en-us/web/css/white-space/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/css/white-space/index.md?plain=1) - Unknown words (stept)
> - [files/en-us/web/html/element/aside/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/html/element/aside/index.md?plain=1) - Unknown words (rainforest)
> - [files/en-us/web/html/element/mark/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/html/element/mark/index.md?plain=1) - Unknown words (rainforest)
> - [files/en-us/web/svg/element/discard/index.md](https://github.com/mdn/content/blob/main/files/en-us/web/svg/element/discard/index.md?plain=1) - Unknown words (Dahlström's)

### A custom CSpell reporter

- As suggested in the https://github.com/streetsidesoftware/cspell/issues/7006 discussion, I've created a custom group-by-file reporter that can be used to reduce the size of the opening post of the weekly spelling check GitHub issues.

OnkarRuikar committed:
ci(cspell-bot): group words by file to reduce log size

OnkarRuikar committed:
need to run yarn install now

OnkarRuikar commented on package.json:
```diff
@@ -54,6 +54,7 @@
     "caporal": "^1.4.0",
     "cli-progress": "^3.12.0",
     "cross-env": "7.0.3",
+    "cspell-group-by-file-reporter": "^1.0.1",
```
All of them should be dev deps because all the dependencies are being used solely for CI or dev preview purposes. Also, we don't pack and ship the content. I did try to change this in my initial days PR, but the changes were denied for no good reason. I think it was "it doesn't matter".

bsmth replied:
>"it doesn't matter"

I guess I agree because "we don't pack and ship the content", it's just muscle memory ¯\_(ツ)_/¯

OnkarRuikar replied:
It's not a case of muscle memory. You don't add linters to main deps. All the linter docs use `--save-dev` in their install instructions. It is a "if all have added to main deps so far, then I also need to add them to the main deps" case.

OnkarRuikar replied:
MOved the linter deps to dev deps.

Josh-Cena commented on .github/workflows/spelling-check-bot.yml:
```diff
@@ -25,10 +25,16 @@ jobs:
           cache: yarn
 
+      - name: Install all yarn packages
+        run: yarn --frozen-lockfile
```
If you need to install anyway, I recommend adding cspell to the deps too so people can spell check locally easily.

OnkarRuikar committed:
add cspell and other linters to dev deps, add lint:typo commands

OnkarRuikar commented:
Made some changes based on the review comments so far:
1. Moved all the linters to dev dependencies.
2. Added CSpell to the dev dependencies. We don't have to use `npx` anymore.
3. Added three new commands to the package.json:
	- `lint:typos` (plain cspell command)
	- `lint:typos-group-by-file` (need this in the GitHub workflow)
	- `lint:typos-words-only` (I need this to fix cspell issues quickly)

OnkarRuikar commented:
Successful run of the changes: https://github.com/OnkarRuikar/content/actions/runs/13804829739/job/38613578932#step:6:68

OnkarRuikar committed:
remove extra info from the output

bsmth commented on files/en-us/mdn/writing_guidelines/writing_style_guide/index.md:
````diff
@@ -373,10 +373,10 @@ Do not use variant spelling.
 ```bash
-npx cspell --no-progress --gitignore --config .vscode/cspell.json "**/*.md"
+yarn lint:typos
 ```
````
Regarding the root cause of this problem, as far as I can see, anything generated from here conforms with our existing cSpell `lorem-ipsum` dictionary: https://www.lipsum.com/feed/html

A good follow-up to this PR is:

* Update the [container scroll-state queries](https://github.com/mdn/content/blob/main/files/en-us/web/css/css_conditional_rules/container_scroll-state_queries/index.md?plain=1) page with passing Lorem text
* Mention this generator (https://www.lipsum.com/feed/html) in our writing guidelines somewhere as a hint for placeholder text.

bsmth replied:
https://github.com/mdn/content/blob/main/.vscode/cspell.json#L23

OnkarRuikar replied:
I held back on submitting typo fixes so that when this PR merges, you'll manually trigger the workflow and we get to test these changes live.

> Update the container scroll-state queries page with passing Lorem text

Agree. There is no point in having other versions of filler texts. We can also recommend the [lorem-ipsum](https://marketplace.visualstudio.com/items?itemName=Tyriar.lorem-ipsum) plugin in our https://github.com/mdn/content/blob/main/.vscode/extensions.json file. I will do the writing guidelines after these issues get resolved.

OnkarRuikar replied:
- the writing guidelines have been updated here: https://github.com/mdn/content/pull/38677

bsmth reviewed (approved):
Thanks, Onkar! LGTM with a comment for when we're going to fix the typos afterwards

OnkarRuikar committed:
update yarn.lock

bsmth committed:
Merge branch 'main' into ci_cspell_bot
